src/daft-minhash/src/windowed.rs

```diff
@@ -0,0 +1,319 @@
+            && self.window_size > self.word_starts.len()
+        {
+            // Yield a window with all words if window_size exceeds the number of words
+            let start = self.word_starts[0];
+            let end = self.s.len();
+            self.current += 1;
+            Some(&self.s[start..end])
+        } else {
+            // No more windows to yield
+            None
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        if self.window_size == 0 {
+            return (0, Some(0));
+        }
+
+        if self.window_size > self.word_starts.len() {
+            if self.word_starts.is_empty() {
+                (0, Some(0))
+            } else {
+                (1, Some(1))
+            }
+        } else {
+            let remaining = self
+                .word_starts
+                .len()
+                .saturating_sub(self.current + self.window_size - 1);
+            (remaining, Some(remaining))
+        }
+    }
+}
+
+impl<'a> ExactSizeIterator for WindowedWords<'a> {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_windowed_words() {
+        let s = "The quick brown fox jumps over the lazy dog";
+        let iter = WindowedWords::new(s, 3);
+        let result: Vec<&str> = iter.collect();
+
+        assert_eq!(
+            result,
+            vec![
+                "The quick brown",
+                "quick brown fox",
+                "brown fox jumps",
+                "fox jumps over",
+                "jumps over the",
+                "over the lazy",
+                "the lazy dog",
+            ]
+        );
+    }
+
+    #[test]
+    fn test_fewer_words_than_window_size() {
+        let s = "Hello world";
+        let iter = WindowedWords::new(s, 3);
+        let result: Vec<&str> = iter.collect();
+
+        assert_eq!(result, vec!["Hello world"]);
+    }
+
+    #[test]
+    fn test_empty_string() {
+        let s = "";
+        let iter = WindowedWords::new(s, 3);
+        let result: Vec<&str> = iter.collect();
+
+        assert_eq!(result, Vec::<&str>::new());
+    }
+
+    #[test]
+    fn test_single_word() {
+        let s = "Hello";
+        let iter = WindowedWords::new(s, 3);
+        let result: Vec<&str> = iter.collect();
+
+        assert_eq!(result, vec!["Hello"]);
+    }
+
+    // currently not supported for performance. see assumptions.
+    // #[test]
+    // fn test_with_extra_whitespace() {
+    //     let s = "  The   quick  brown   ";
+    //     let iter = WindowedWords::new(s, 2);
+    //     let result: Vec<&str> = iter.collect();
+    //
+    //     assert_eq!(result, vec!["The   quick", "quick  brown"]);
+    // }
+
+    #[test]
+    fn test_large_window_size() {
+        let s = "One two three";
+        let iter = WindowedWords::new(s, 5);
+        let result: Vec<&str> = iter.collect();
+
+        assert_eq!(result, vec!["One two three"]);
+    }
+
+    // currently not supported for performance. see assumptions.
```

Comment:
This is something some other projects like bigcode support and we didn't. It could be interesting to have our windower to give us a iterator of `&str` that we can then feed our hasher.

so something like `"Hello    world"` would yield `{"hello", " ", "world"}` via the iterator which can be consumed by the hasher.

Reply:
down to talk about this more

Reply:
marking this as resolved for now